Userland/DevTools/UserspaceEmulator/SoftFPU.cpp

```diff
@@ -0,0 +1,1744 @@
+        double f64 = fpu_convert_checked<double>(fpu_get(0));
+        if (fpu_is_set(0))
+            insn.modrm().write64(m_cpu, insn, shadow_wrap_as_initialized(bit_cast<u64>(f64)));
+        else
+            insn.modrm().write64(m_cpu, insn, ValueWithShadow<u64>(bit_cast<u64>(f64), 0ULL));
+    }
+}
+
+void SoftFPU::FSTP_RM32(const X86::Instruction& insn)
+{
+    FST_RM32(insn);
+    fpu_pop();
+}
+void SoftFPU::FSTP_RM64(const X86::Instruction& insn)
+{
+    FST_RM64(insn);
+    fpu_pop();
+}
+void SoftFPU::FSTP_RM80(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        fpu_set(insn.modrm().register_index(), fpu_get(0));
+        fpu_pop();
+    } else {
+        // FIXME: Respect more shadow values
+        // long doubles can be up to 128 bits wide in memory for reasons (alignment) and only uses 80 bits of precision
+        // gcc uses 12 byte in 32 bit and 16 byte in 64 bit mode
+        // due to only 10 bytes being used, we just write these 10 into memory
+        // We have to do .bytes().data() to get around static type analysis
+        ValueWithShadow<u128> f80 { 0u, 0u };
+        u128 value {};
+        f80 = insn.modrm().read128(m_cpu, insn);
+        *(long double*)value.bytes().data() = fpu_pop();
+        memcpy(f80.value().bytes().data(), &value, 10); // copy
+        memset(f80.shadow().bytes().data(), 0x01, 10);  // mark as initialized
+        insn.modrm().write128(m_cpu, insn, f80);
+    }
+}
+
+void SoftFPU::FILD_RM16(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto m16int = insn.modrm().read16(m_cpu, insn);
+    warn_if_uninitialized(m16int, "int16 loaded as float");
+
+    fpu_push(static_cast<long double>(static_cast<i16>(m16int.value())));
+}
+void SoftFPU::FILD_RM32(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto m32int = insn.modrm().read32(m_cpu, insn);
+    warn_if_uninitialized(m32int, "int32 loaded as float");
+
+    fpu_push(static_cast<long double>(static_cast<i32>(m32int.value())));
+}
+void SoftFPU::FILD_RM64(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto m64int = insn.modrm().read64(m_cpu, insn);
+    warn_if_uninitialized(m64int, "int64 loaded as float");
+
+    fpu_push(static_cast<long double>(static_cast<i64>(m64int.value())));
+}
+
+void SoftFPU::FIST_RM16(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto f = fpu_get(0);
+    set_c1(0);
+    auto int16 = fpu_round_checked<i16>(f);
+
+    // FIXME: Respect shadow values
+    insn.modrm().write16(m_cpu, insn, shadow_wrap_as_initialized(bit_cast<u16>(int16)));
+}
+void SoftFPU::FIST_RM32(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto f = fpu_get(0);
+    set_c1(0);
+    auto int32 = fpu_round_checked<i32>(f);
+    // FIXME: Respect shadow values
+    insn.modrm().write32(m_cpu, insn, shadow_wrap_as_initialized(bit_cast<u32>(int32)));
+}
+
```

Comment:
No FIST_RM64?

Reply:
that one always pops too
https://www.felixcloutier.com/x86/fist:fistp